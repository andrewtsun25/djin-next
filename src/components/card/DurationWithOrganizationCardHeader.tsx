import { Avatar, CardHeader } from "@mui/material";
import React from "react";

import { getTimeIntervalAsString } from "../../util/date";
import { CurrentBadge } from "./styled";

interface DurationWithOrganizationCardHeaderProps {
  title: string;
  subtitle: string;
  startDate: Date;
  endDate?: Date | null;
  logoUrl?: string;
}

const CURRENT_BADGE_TEST_ID = "Current Badge";

export const DurationWithOrganizationCardHeader: React.FC<
  DurationWithOrganizationCardHeaderProps
> = ({
  title,
  subtitle,
  startDate,
  endDate,
  logoUrl,
}: DurationWithOrganizationCardHeaderProps) => {
  const subheader = `${getTimeIntervalAsString(startDate, endDate)}: ${subtitle}`;
  const baseAvatar: React.ReactNode = <Avatar alt={title} src={logoUrl} />;
  const avatar: React.ReactNode = endDate ? (
    baseAvatar
  ) : (
    <CurrentBadge
      overlap="circular"
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "right",
      }}
      variant="dot"
      data-testid={CURRENT_BADGE_TEST_ID}
    >
      {baseAvatar}
    </CurrentBadge>
  );
  return <CardHeader title={title} subheader={subheader} avatar={avatar} />;
};

export { CURRENT_BADGE_TEST_ID };
export type { DurationWithOrganizationCardHeaderProps };
