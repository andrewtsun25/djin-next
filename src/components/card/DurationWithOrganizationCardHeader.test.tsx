import { render } from "@testing-library/react";

import DurationWithOrganizationCardHeader, {
  CURRENT_BADGE_TEST_ID,
} from "./DurationWithOrganizationCardHeader";

describe("DurationWithOrganizationCardHeader", () => {
  const title = "Title";
  const subtitle = "Subtitle";
  const startDate = new Date("2022-12-25T12:00:00Z");
  const endDate = new Date("2023-01-01T12:00:00Z");
  const logoUrl = "https://via.placeholder.com/300.png";

  it("renders", () => {
    const durationWithOrganizationCardHeader = render(
      <DurationWithOrganizationCardHeader
        title={title}
        subtitle={subtitle}
        startDate={startDate}
        endDate={endDate}
        logoUrl={logoUrl}
      />
    );
    expect(durationWithOrganizationCardHeader).toMatchSnapshot();
  });

  it("renders the end date if the end date is provided", () => {
    const { getByText } = render(
      <DurationWithOrganizationCardHeader
        title={title}
        subtitle={subtitle}
        startDate={startDate}
        endDate={endDate}
        logoUrl={logoUrl}
      />
    );
    const datesAndSubtitle = getByText("Dec 25, 2022 - Jan 1, 2023: Subtitle");
    expect(datesAndSubtitle).not.toBeNull();
  });

  it("renders the avatar with the image at the logo url if the logo url is provided", () => {
    const { getByAltText } = render(
      <DurationWithOrganizationCardHeader
        title={title}
        subtitle={subtitle}
        startDate={startDate}
        endDate={endDate}
        logoUrl={logoUrl}
      />
    );
    const logo = getByAltText(title);
    expect(logo).toBeVisible();
  });

  it("renders present for the end date if the end date is not provided", () => {
    const { getByText, getByTestId } = render(
      <DurationWithOrganizationCardHeader
        title={title}
        subtitle={subtitle}
        startDate={startDate}
        logoUrl={logoUrl}
      />
    );
    const datesAndSubtitle = getByText("Dec 25, 2022 - Present: Subtitle");
    expect(datesAndSubtitle).not.toBeNull();
    const currentBadge = getByTestId(CURRENT_BADGE_TEST_ID);
    expect(currentBadge).not.toBeNull();
  });

  it("renders the person icon in the Avatar if logoUrl is not provided", () => {
    const { getByTestId } = render(
      <DurationWithOrganizationCardHeader
        title={title}
        subtitle={subtitle}
        startDate={startDate}
        endDate={endDate}
      />
    );
    const avatarAltText = getByTestId("PersonIcon");
    expect(avatarAltText).not.toBeNull();
  });
});
