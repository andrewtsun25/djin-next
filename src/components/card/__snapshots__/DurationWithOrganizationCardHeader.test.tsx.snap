// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DurationWithOrganizationCardHeader renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
          >
            <img
              alt="Title"
              class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
              src="https://via.placeholder.com/300.png"
            />
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
          >
            Title
          </span>
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
          >
            Dec 25, 2022 - Jan 1, 2023: Subtitle
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
    >
      <div
        class="MuiCardHeader-avatar css-1ssile9-MuiCardHeader-avatar"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1wlk0hk-MuiAvatar-root"
        >
          <img
            alt="Title"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="https://via.placeholder.com/300.png"
          />
        </div>
      </div>
      <div
        class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
      >
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-et1ao3-MuiTypography-root"
        >
          Title
        </span>
        <span
          class="MuiTypography-root MuiTypography-body2 MuiCardHeader-subheader css-83ijpv-MuiTypography-root"
        >
          Dec 25, 2022 - Jan 1, 2023: Subtitle
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
